### ISSUE TITLE
Null ptr check doesn't catch null reference to ZSTs

### ISSUE THREAD

**Issue by zjp-CN** [rust-lang/rust#136568]
As https://github.com/rust-lang/rust/pull/134424 recently landed, for the following code, I'd expect the check will catch null reference as Miri does, but actually not:

```rust
// Based on the description of PR#134424, but using a ptr to ZST
let ptr: *const () = std::ptr::null();
let _val: &() = unsafe { &*ptr };
```

As per https://github.com/rust-lang/unsafe-code-guidelines/issues/472 , access to a ptr to ZST is not UB, so  `*ptr` is ok.

```rust
$ cargo run
# **No panic** like "null pointer dereference occured"
```

But it could be better to catch null reference `&*ptr`:

```rust
$ cargo miri run
error: Undefined Behavior: constructing invalid value: encountered a null reference
 --> src/main.rs:3:30
  |
3 |     let _val: &() = unsafe { &*ptr };
  |                              ^^^^^ constructing invalid value: encountered a null reference
  |
  = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
  = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
```

Not sure if I put up a real issue. I just thought the check would act on null reference, since in some cases [the term pointer may refer to a reference as well](https://doc.rust-lang.org/reference/types/pointer.html).

I guess we need another check for null reference, because null ptr is not harmful, dereferencing to it is harmful, while it doesn't need dereferencing for null reference to be UB.

**Comment by workingjubilee** [comment]
@zjp-CN This cannot be checked by library code.

**Comment by workingjubilee** [comment]
Hm, the pass does insert check for borrows... odd.

**Comment by compiler-errors** [comment]
We don't insert null checks for places of ZSTs regardless of how those places are being used. I opened #136601 to ensure that we enforce this for borrows *even* for ZSTs.
